api/src/modules/impact/scenario-comparison.service.ts

```diff
@@ -0,0 +1,829 @@
+     * has canceled and replaced impacts  + cancelled and replaced tonnes grouped by scenarioId -
+     * in the properties scenarioImpacts and scenarioTonnes
+     */
+    impactData.forEach((impactData: ImpactTableData) => {
+      const existingScenarioData:
+        | ScenariosComparisonImpactTableData
+        | undefined = dataForScenarioImpactTable.find((el: ImpactTableData) => {
+        return el.name === impactData.name && el.year === impactData.year;
+      });
+
+      if (existingScenarioData) {
+        dataForScenarioImpactTable[
+          dataForScenarioImpactTable.indexOf(existingScenarioData)
+        ].scenariosImpacts.push({
+          scenarioId: impactData.scenarioId as string,
+          ...(impactData.typeByIntervention ===
+            SOURCING_LOCATION_TYPE_BY_INTERVENTION.REPLACING && {
+            newImpact: impactData.impact,
+          }),
+          ...(impactData.typeByIntervention ===
+            SOURCING_LOCATION_TYPE_BY_INTERVENTION.CANCELED && {
+            canceledImpact: impactData.impact,
+          }),
+        });
+
+        dataForScenarioImpactTable[
+          dataForScenarioImpactTable.indexOf(existingScenarioData)
+        ].scenariosTonnes.push({
+          scenarioId: impactData.scenarioId as string,
+          ...(impactData.typeByIntervention ===
+            SOURCING_LOCATION_TYPE_BY_INTERVENTION.REPLACING && {
+            newTonnage: Number(impactData.tonnes),
+          }),
+          ...(impactData.typeByIntervention ===
+            SOURCING_LOCATION_TYPE_BY_INTERVENTION.CANCELED && {
+            canceledTonnage: Number(impactData.tonnes),
+          }),
+        });
+      } else {
+        const newComparisonData: ScenariosComparisonImpactTableData = {
+          ...impactData,
+          scenariosImpacts: [
+            {
+              scenarioId: impactData.scenarioId as string,
+              ...(impactData.typeByIntervention ===
+                SOURCING_LOCATION_TYPE_BY_INTERVENTION.REPLACING && {
+                newImpact: impactData.impact,
+              }),
+              ...(impactData.typeByIntervention ===
+                SOURCING_LOCATION_TYPE_BY_INTERVENTION.CANCELED && {
+                canceledImpact: impactData.impact,
+              }),
+            },
+          ],
+          scenariosTonnes: [
+            {
+              scenarioId: impactData.scenarioId as string,
+              ...(impactData.typeByIntervention ===
+                SOURCING_LOCATION_TYPE_BY_INTERVENTION.REPLACING && {
+                newTonnage: Number(impactData.tonnes),
+              }),
+              ...(impactData.typeByIntervention ===
+                SOURCING_LOCATION_TYPE_BY_INTERVENTION.CANCELED && {
+                canceledTonnage: Number(impactData.tonnes),
+              }),
+            },
+          ],
+        };
+        dataForScenarioImpactTable.push(newComparisonData);
+      }
+    });
```

Comment:
This section has several parts of code that are very similar; maybe it can be extracted to a small function to make it more readable?